library/java/org/chromium/net/impl/CronetBidirectionalState.java

```diff
@@ -0,0 +1,431 @@
+                                                           : NextAction.INVOKE_ON_CANCELED;
+        break;
+
+      case Event.ON_ERROR:
+        nextState |= State.DONE | State.FAILED;
+        nextAction = ((originalState & State.FAILED) != 0) ? NextAction.INVOKE_ON_FAILED
+                                                           : NextAction.INVOKE_ON_ERROR_RECEIVED;
+        break;
+
+      case Event.LAST_WRITE_COMPLETED:
+        assert (originalState & State.WRITE_DONE) == 0;
+        nextState |= State.WRITE_DONE;
+        // FOLLOW THROUGH
+      case Event.WRITE_COMPLETED:
+        nextAction = NextAction.INVOKE_ON_WRITE_COMPLETED_CALLBACK;
+        break;
+
+      case Event.READY_TO_FLUSH:
+        if ((originalState & State.WAITING_FOR_FLUSH) == 0) {
+          nextAction = NextAction.CARRY_ON;
+        } else {
+          nextState &= ~State.WAITING_FOR_FLUSH;
+          nextState |= State.WRITING;
+          nextAction = NextAction.SEND_DATA;
+        }
+        break;
+
+      case Event.FLUSH_DATA_COMPLETED:
+        nextState |= State.WAITING_FOR_FLUSH;
+        // FOLLOW THROUGH
+      case Event.LAST_FLUSH_DATA_COMPLETED:
+        assert (originalState & State.WRITING) != 0;
+        assert (originalState & State.WAITING_FOR_FLUSH) == 0;
+        nextState &= ~State.WRITING;
+        nextAction = NextAction.CARRY_ON;
+        break;
+
+      case Event.READY_TO_READ:
+        assert (originalState & State.WAITING_FOR_READ) == 0;
+        assert (originalState & State.READING) == 0;
+        nextState |= State.READING;
+        nextAction = (originalState & State.ON_HEADER_RECEIVED) == 0 ? NextAction.CARRY_ON
+                     : (originalState & State.END_STREAM_READ) == 0
+                         ? NextAction.READ
+                         : NextAction.INVOKE_ON_READ_COMPLETED;
+        break;
+
+      case Event.READ_COMPLETED:
+        assert (originalState & State.READING) != 0;
+        nextState &= ~State.READING;
+        nextState |= State.WAITING_FOR_READ;
+        nextAction = NextAction.INVOKE_ON_READ_COMPLETED_CALLBACK;
+        break;
+
+      case Event.LAST_READ_COMPLETED:
+        assert (originalState & State.READ_DONE) == 0;
+        nextState &= ~State.READING;
+        nextState |= State.READ_DONE;
+        nextAction = NextAction.INVOKE_ON_READ_COMPLETED_CALLBACK;
+        break;
+
+      case Event.READY_TO_FINISH:
+        if ((originalState & State.ON_COMPLETE_RECEIVED) != 0 &&
+            (originalState & State.READ_DONE) != 0 && (originalState & State.WRITE_DONE) != 0) {
+          nextState |= State.DONE;
+          nextAction = NextAction.INVOKE_ON_SUCCEEDED;
+        } else {
+          nextAction = NextAction.CARRY_ON;
+        }
+        break;
+
+      default:
+        throw new AssertionError("switch is exhaustive");
+      }
+
+      if (mState.compareAndSet(originalState, nextState)) {
+        return nextAction;
+      }
```

Comment:
There is a LOT of logic in this loop, not surprisingly. I'm not quite sure how to verify that all the transitions are correct. Is there a Cronet state transition diagram somewhere?

Reply:
The complexity lies in the fact that EM and Cronet offer a different API, and because with CAS incurs some internal operations. The exact diagram is quite big, unwieldy. The best short term is to explain to you the general lines.

At this point, the only think that can give you a little not of confidence are the Cronet tests: they are the same as the original ones. Furthermore, I chased for weeks all the race conditions. I ran the test in batches of 500, many times. Good news, the tests are quite good at finding issues under heavy contention (all my M1 CPUs were running hot...)

Reply:
As you say, I suspect that this is "correct" and that the testing you did gives good confidence for that. But I have no idea how to look at this code and understand if it is correct. For example, this case:
```
      case Event.READY_TO_FINISH:
        if ((originalState & State.ON_COMPLETE_RECEIVED) != 0 &&
            (originalState & State.READ_DONE) != 0 && (originalState & State.WRITE_DONE) != 0) {
          nextState |= State.DONE;
          nextAction = NextAction.INVOKE_ON_SUCCEEDED;
        } else {
          nextAction = NextAction.CARRY_ON;
        }
        break;
```

It seems like this suggests that we're done when we've finished reading AND we've finished writing (and we get ON_COMPLETE_RECEIVED). That sounds plausible. But are there other things that should have finished too? Probably not :) Are there perhaps two different state machines that we could consider: a Read state machine and a Write state machine? (This matches some of the state machines in Chrome net code and matches the QUIC Stream state machine in RFC 9000 so maybe it makes sense here). If changes to the Read state machine do not affect the Write state machine, this might be a way to simplify the complexity. But maybe this is not the right mental model?

Reply:
Discussed in person. Status quo for the moment.